Store union of hashes for block witness

carver opened the pull request:
### What was wrong?

Fixes #2105 

### How was it fixed?

Store union of the witness hashes, when witness metadata comes in for the same block.

### To-Do

- [x] ~Test where two persists interleave their database accesses, to force the race condition that one witness overwrites the other~
- [x] ~Add a lock to pass ^ test~
- [ ] Clean up commit history

[//]: # (For important changes that should go into the release notes please add a newsfragment file as explained here: https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

[//]: # (See: https://trinity-client.readthedocs.io/en/latest/contributing.html#pull-requests)
- [x] ~Add entry to the [release notes](https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)~

#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](https://i.pinimg.com/originals/b2/fb/19/b2fb195935593827ae6dd157139c1208.jpg)

gsalgado commented:
Sorry, I keep forgetting to review this. Will do so tomorrow, but just thought I'd mention I found a bug that causes `persist_witness_hashes()` to crash with a KeyError when we call it multiple times with the same hash. This test exhibits it: https://github.com/ethereum/trinity/commit/d0b33b51f19a3d84d32bf47440dcbd4b759cdfb2

Is that bug present in this branch as well?

carver commented:
Yup, looks like that bug is fixed in this branch. I added the test. It's similar to another test I added, but seems good to have another explicit one. :+1:

gsalgado commented on tests/core/p2p-proto/test_wit_db.py:
```diff
@@ -42,3 +45,66 @@ async def test_witness_for_recent_blocks():
+    """
+    Repeated blocks should not consume more slots in the limited history of block witnesses
+    """
+    wit_db = AsyncWitnessDB(AtomicDB())
```
AsyncWitnessDB was changed recently to wrap a regular `DatabaseAPI` instead of `AtomicDatabaseAPI`, so we could use a `MemoryDB()` here instead

gsalgado reviewed (approved):
LGTM
